### PR TITLE
Minor cleanup in visibleFunctions et al.

### PR DESCRIPTION
**Pull request by vasslitvinov**
Reduces calls to getModule() and getFunction() and improves code clarity.

Removes or conditional-izes a couple of calls to getModule() and getFunction()
in visibleFunctions.cpp where that does not affect business logic.
This was triggered by an assertion failure on my w.i.p. CG branch.

Replaces some code in getVisibleMethodsImpl() with calls to
getVisibleFnsInstantiationPt() and getVisibleFnsShowBlock(),
sharing these with getVisibleFunctionsImpl().

Moves the check for rootModule->block to be the first thing in
getVisibleMethodsImpl() and getVisibleFunctionsImpl(), to
avoid unnecessary work.

Reworks some comments for clarity or conciseness.
The changes in createTaskFunctions.cpp were triggered
by the discovery that tiMarkHost was the only function
with a defPoint directly in rootModule.

### CODE REVIEW

**Comment by lydia-duncan** [comment]
> Replaces some code in getVisibleMethodsImpl() with calls to
getVisibleFnsInstantiationPt() and getVisibleFnsShowBlock(),
sharing these with getVisibleFunctionsImpl().

Heh, I also have a branch that does this, guess I should have separated it out and merged it earlier

File: compiler/resolution/visibleFunctions.cpp [hunk 1]
```diff
@@ -564,7 +534,7 @@ static void getVisibleMethodsImpl(const char* name, CallExpr* call,
       }
     }
 
-    if (block != rootModule->block) {
+    {
```
Comment (lydia-duncan) [inline compiler/resolution/visibleFunctions.cpp:537]: > Moves the check for rootModule->block to be the first thing in
getVisibleMethodsImpl() and getVisibleFunctionsImpl(), to
safe unnecessary work.

This line and the addition on line 465 have to do with this part of your PR message, right?  I don't think this is doing what you think it's doing, if so.  The check here is to prevent us from getting the parent scope of a scope that doesn't have a parent scope.  rootModule has included things in the past - we've placed some code the compiler generates in there, if I remember right.  Have you gone through the contents of the rootModule carefully to be certain that there isn't anything in it any more?  Otherwise, I would be worried we're skipping symbols that should be considered.

At the very least, please don't leave random `{}` lying around, it clutters the code.

File: compiler/resolution/visibleFunctions.cpp [hunk 2]
```diff
@@ -645,21 +615,15 @@ void getVisibleFunctions(const char*      name,
...
 
-  if (block->parentExpr != NULL) {
-    // not a module or function level block
-  } else if (inMod && block == inMod->block) {
-    // module-level block
-  } else if (inFn != NULL) {
-    // TODO - probably remove this assert
-    INT_ASSERT(block->parentSymbol == inFn ||
-               isArgSymbol(block->parentSymbol) ||
-               isShadowVarSymbol(block->parentSymbol));
```
Comment (lydia-duncan) [inline compiler/resolution/visibleFunctions.cpp:661]: I'm assuming this is the assert that was causing problems for you?  The TODO here does make me inclined to okay its removal, but I would be slacking in my review if I didn't at least double check why we think it is okay to remove it.

File: compiler/resolution/visibleFunctions.cpp [hunk 3]
```diff
@@ -869,7 +837,7 @@ static void getVisibleFunctionsImpl(const char*       name,
 
   // Need to continue going up in case our parent scopes also had private
   // uses that were skipped.
-  if (block != rootModule->block) {
+  {
```
Comment (lydia-duncan) [inline compiler/resolution/visibleFunctions.cpp:840]: Ditto here about the rootModule block stuff

File: compiler/resolution/visibleFunctions.cpp [hunk 4]
```diff
@@ -954,7 +922,7 @@ BlockStmt* getVisibilityScope(Expr* expr) {
   while (cur != NULL) {
     // Pretend that ArgSymbols are in the function's body
     // (which is reasonable since functions cannot be defined
-    //  within an ArgSymbol).
+    //  within an ArgSymbol), so we consult POIs as needed.
```
Comment (lydia-duncan) [inline compiler/resolution/visibleFunctions.cpp:925]: I'm not sure I understand what this addition means

**Review by lydia-duncan — COMMENTED** [review]
Just a few questions

File: compiler/resolution/visibleFunctions.cpp (hunk 1 above)
Comment (vasslitvinov) [reply compiler/resolution/visibleFunctions.cpp:537]: Actually, I did. That's where the changes in createTaskFunction.cpp came from: because tiMarkHost is the only thing in rootModule right now.

I don't know how valuable it is to place functions in rootModule. Instead, we could place them in `theProgram`.

I will remove the {}.

File: compiler/resolution/visibleFunctions.cpp (hunk 2 above)
Comment (vasslitvinov) [reply compiler/resolution/visibleFunctions.cpp:661]: Correct. Good point.

I made the change not because it was logically identical, but on semantics grounds. We care about the point of instantiation only for a function. (Even if it is a method on a generic class.) Therefore getVisibleFnsInstantiationPt() should return the POI only for its top-level blocks, such as `body` or `where`. So I wrote the code that does that, instead of going roundabout it.

File: compiler/resolution/visibleFunctions.cpp (hunk 3 above)
Comment (vasslitvinov) [reply compiler/resolution/visibleFunctions.cpp:840]: Right, same situation.

File: compiler/resolution/visibleFunctions.cpp (hunk 4 above)
Comment (vasslitvinov) [reply compiler/resolution/visibleFunctions.cpp:925]: I meant to say that by returning `fn->body` in this case, we enable getVisibleFunctions() to visit the function's POI. Which is relied upon in the test mentioned in the next comment line, `test/functions/default-arguments/default-argument-generic.chpl`

Comment (vasslitvinov) [reply compiler/resolution/visibleFunctions.cpp:925]: P.S. Let me know if you have a better way to phrase it.

**Comment by vasslitvinov** [comment]
@lydia-duncan I responded to your requests. Please let me know what else is needed for merging.

File: compiler/resolution/visibleFunctions.cpp (hunk 4 above)
Comment (lydia-duncan) [reply compiler/resolution/visibleFunctions.cpp:915]: So I think what you're saying is that returning the function's body when getting the visibility scope for an ArgSymbol allows us to resolve things that ArgSymbol relies upon, if they are defined at the point of instantiation?  E.g.
```
module A {
  proc someFunc(type t, x = otherFunc()) { ... }
}
module B {
  use A;
  proc otherFunc() { ... }
  proc caller() {
    someFunc(int); // Where this is the point of instantiation, and that allows us to find B.otherFunc()?
  }
}
```

Comment (lydia-duncan) [reply compiler/resolution/visibleFunctions.cpp:915]: If that's the case, I think I would rephrase this to: "Pretend that ArgSymbols are in the function's body (which is reasonable since functions cannot be defined within an ArgSymbol), so that we can consult the function's POIs when resolving the ArgSymbol."

File: compiler/resolution/visibleFunctions.cpp (hunk 2 above)
Comment (lydia-duncan) [reply compiler/resolution/visibleFunctions.cpp:661]: I forget, are we allowed to define nested functions at arbitrary points within a function?  Or do they have to be at the top level of the function body?

E.g. can I write this?
```chapel
proc foo(...) {
  if blah {
    proc innerFoo() { ... }
    innerFoo();
  }
}
```

Comment (lydia-duncan) [reply compiler/resolution/visibleFunctions.cpp:661]: I don't think we can define classes in those bodies any more, so maybe it doesn't matter

**Comment by lydia-duncan** [comment]
I think just update the comment with the new phrasing if you think it's clearer (or suggest an alternative if you think of something better).  Other than that, I think this is good to go

File: compiler/resolution/visibleFunctions.cpp (hunk 2 above)
Comment (vasslitvinov) [reply compiler/resolution/visibleFunctions.cpp:661]: Yes, a nested function can be defined in a nested scope within the enclosing function. The relevant piece here is that a nested scope, such as the then-block of a conditional, does not have a POI. getVisibleFnsInstantiationPt() rightfully returns NULL in this case.

Re classes: a method on a generic class has its own POI when instantiated, it does not rely on the POI of the class. I think that classes don't have their own POIs.

Comment (lydia-duncan) [reply compiler/resolution/visibleFunctions.cpp:661]: > Yes, a nested function can be defined in a nested scope within the enclosing function. The relevant piece here is that a nested scope, such as the then-block of a conditional, does not have a POI. getVisibleFnsInstantiationPt() rightfully returns NULL in this case.

That seems problematic, though.  If we've written a nested function that fulfills what a generic function is looking for, would we not be able to use the nested function?  Or would two calls at different scopes in the same function use the same instantiation, even if the first was able to see the nested function and the second wasn't?  E.g.

```chapel
proc genericFunc(param p) {
  poiCall();
}
proc other() {
  if blah {
    proc poiCall() { ... }
    genericFunc(1); // Should this instantiation use the definition above?
  }
  // I don't think this call should reuse the instantiation
  // since it's not in the scope of the poiCall definition
  genericFunc(1);
}
```

Comment (vasslitvinov) [reply compiler/resolution/visibleFunctions.cpp:661]: Good example. To foreshadow the following discussion, the compiler does the right thing.

Let us see what is happening during resolution of the above code.

(A) When resolving either call to genericFunc(), POI does not come into play. getVisibleFunctionsImpl() goes up the lexical scopes until it reaches the body of the implicit module, where it finds the definition of genericFunc().

(B) The POI becomes relevant when resolving the call to poiCall() within genericFunc(). There, getVisibleFnsInstantiationPt() returns NULL whenever its argument is not the top-level block of genericFunc's body. For example, if the call to poiCall() were nested within a conditional:

```chpl
proc genericFunc(param p) {
  if blah {
    poiCall();
  }
}
```

then the first time getVisibleFnsInstantiationPt() is invoked it would return NULL. This is the correct behavior.

Meanwhile getVisibleFunctionsImpl() continues marching up the lexical scopes. Once it reaches genericFunc's top body block, getVisibleFnsInstantiationPt() wakes up and returns genericFunc's POI.

That POI itself can be nested or not. However, it will be the job of getVisibleFunctionsImpl() to traverse up the lexical scopes starting from whatever POI it got.

In the above example, the POI for the first call to genericFunc() is nested within the conditional whereas the second one is not. Either way, getVisibleFnsInstantiationPt() is not affected by this difference because it returns them without examining them.

<details>
<summary>A compilable example</summary>

```chpl
config var blah = true;

proc genericFunc(param p) {
  if blah {
    poiCall();
  }
}
proc other() {
  if blah {
    proc poiCall() { writeln("inner"); }
    genericFunc(1);
  }
  proc poiCall() { writeln("outer"); } // commenting this out would give an "unresolved" error
  genericFunc(1);
}

proc main {
  other();
}
```

</details>

Comment (lydia-duncan) [reply compiler/resolution/visibleFunctions.cpp:661]: Okay, sweet!  Thanks for walking through it
